### PR TITLE
SslHandler unwrap out of order promise/event notificaiton

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
SslHandler#decode methods catch any exceptions and attempt to wrap
before shutting down the engine. The intention is to write any alerts
which the engine may have pending. However the wrap process may also
attempt to write user data, and may also complete the associated
promises. If this is the case, and a promise listener closes the channel
then SslHandler may later propagate a SslHandshakeCompletionEvent user
event through the pipeline. Since the channel has already been closed
the user may no longer be paying attention to user events.

Modifications:
- Sslhandler#decode should first fail the associated handshake promise
and propagate the SslHandshakeCompletionEvent before attempting to wrap

Result:
Fixes https://github.com/netty/netty/issues/7639

### CODE REVIEW

**Comment by Scottmitch** [comment]
@carl-mastrangelo - can you verify this PR fixes the original issue?

**Comment by carl-mastrangelo** [comment]
@Scottmitch Unfortunately I can't test it out.  Between 4.1.17 and 4.1.20 Netty stopped working for gRPC (tests fail plus https://github.com/grpc/grpc-java/issues/4025 ).   I haven't had a chance to update gRPC for this usage.   (too many Netty PRs in my backlog ;) )

File: handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java [hunk 1]
```diff
@@ -431,6 +437,97 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
...
+
+            SslCompletionEvent evt = (SslCompletionEvent) events.take();
+            assertTrue(evt instanceof SslHandshakeCompletionEvent);
+            assertThat(evt.cause(), is(instanceOf(SSLException.class)));
+
+            ChannelFuture future = (ChannelFuture) events.take();
+            assertThat(future.cause(), is(instanceOf(SSLException.class)));
+
+            serverChannel.close();
+            clientChannel.close();
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java]: @Scottmitch nit: also null these out so we not try to close again in the finally block.

Comment (normanmaurer) [reply handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java]: Also you may want to call `sync()`, not needed but still not bad I think

File: handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java [hunk 2]
```diff
@@ -431,6 +437,97 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
...
+                .channel(LocalServerChannel.class)
+                .childHandler(new ChannelInitializer<Channel>() {
+                  @Override
+                  protected void initChannel(Channel ch) {
+                      ch.pipeline().addLast(sslServerCtx.newHandler(ch.alloc()));
+                      ch.pipeline().addLast(new ChannelInboundHandlerAdapter() {
+                          @Override
+                          public void channelActive(ChannelHandlerContext ctx) {
+                              ByteBuf buf = ctx.alloc().buffer(10);
+                              buf.writerIndex(buf.readerIndex() + 10);
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java]: nit: you could also just call `buf.writeZero(10)`.

Comment (Scottmitch) [reply handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java]: done

File: handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java (hunk 1 above)
Comment (Scottmitch) [reply handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java]: done

File: buffer/src/test/java/io/netty/buffer/AbstractByteBufTest.java [hunk 3]
```diff
@@ -19,6 +19,7 @@
 import io.netty.util.CharsetUtil;
 import io.netty.util.IllegalReferenceCountException;
 import io.netty.util.internal.PlatformDependent;
+import javafx.application.Platform;
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/AbstractByteBufTest.java]: Remove ?

File: buffer/src/test/java/io/netty/buffer/AbstractByteBufTest.java [hunk 4]
```diff
@@ -115,6 +116,9 @@ public void dispose() {
 
     @Test
     public void comparableInterfaceNotViolated() {
+        ByteBuffer bb = ByteBuffer.allocateDirect(10);
+        ByteBuffer bbro = bb.asReadOnlyBuffer();
+        long f = PlatformDependent.directBufferAddress(bbro);
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/AbstractByteBufTest.java]: Actually remove the change ?

**Comment by ejona86** [comment]
@Scottmitch, I back-ported the patch to 4.1.17 and can confirm it fixes the issue.

File: handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java [hunk 5]
```diff
@@ -431,6 +437,99 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
...
+            assertThat(evt.cause(), is(instanceOf(ClosedChannelException.class)));
+            assertTrue(events.isEmpty());
+        } finally {
+            if (serverChannel != null) {
+                serverChannel.close();
+            }
+            if (clientChannel != null) {
+                clientChannel.close();
+            }
+            group.shutdownGracefully();
```
Comment (normanmaurer) [inline handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java]: Nit: call ssc.delete()

Comment (Scottmitch) [reply handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java]: We don't consistently do this in tests. Lets do this in a followup PR if we think it is necessary.

**Comment by Scottmitch** [comment]
4.1 (978a46cc0a) 4.0 (94e55692d4)

**Pull request closed without merging**